k2/csrc/determinize.h

```diff
@@ -371,20 +372,20 @@ class DetState;
 
 template <class TracebackState>
 struct DetStateCompare {
-  bool operator()(const std::unique_ptr<DetState<TracebackState>> &a,
-                  const std::unique_ptr<DetState<TracebackState>> &b) {
+  bool operator()(const std::shared_ptr<DetState<TracebackState>> &a,
+                  const std::shared_ptr<DetState<TracebackState>> &b) {
     return a->forward_backward_prob < b->forward_backward_prob;
   }
 };
 // Priority queue template arguments:
-//   item queued = unique_ptr<DetState> (using pointer equality as comparison)
-//   container type = vector<unique_ptr<DetState> >
+//   item queued = shared_ptr<DetState> (using pointer equality as comparison)
+//   container type = vector<shared_ptr<DetState> >
 //   less-than operator = DetStateCompare (which compares the
 //   forward_backward_prob).
 template <class TracebackState>
 using DetStatePriorityQueue =
-    std::priority_queue<std::unique_ptr<DetState<TracebackState>>,
-                        std::vector<std::unique_ptr<DetState<TracebackState>>>,
+    std::priority_queue<std::shared_ptr<DetState<TracebackState>>,
+                        std::vector<std::shared_ptr<DetState<TracebackState>>>,
```

Comment:
What was the problem with unique_ptr here?  I made it unique_ptr as it's more lightweight than shared_ptr (and this is only ever owned in one place, I believed.)

Reply:
`priority_queue.top()` returns a const reference so we cannnot call `std::move()` on its returned value to construct any `unqiue_ptr` or `shared_ptr`.

Reply:
see here 
https://github.com/danpovey/k2/blob/75c4cd1b1fdd2a20009997c269b905b438575cdb/k2/csrc/determinize.h#L818

Reply:
I see; shared_ptr is fine.